### PR TITLE
Flag encoder versions

### PR DESCRIPTION
**Pull request by andreaswolf**
This adds version numbers to all encoder classes. The numbers are stored with the graph and checked on reloading.

### CODE REVIEW

**Comment by karussell** [comment]
Thanks!

Looks good! Two minor things: would you use PMap instead of the new properties class? And would you rebase the changes into one commit without the [BRACKET] notation? (it is just uncommon for graphhopper, and maybe we introduce something for the different modules [web/mobile/...] but step by step :))

**Comment by andreaswolf** [comment]
I adjusted to your comments. I did not squash the patch to make the changes between versions 1 and 2 clear. If you think the feature is worth merging, I’ll squash the commits to one.

File: core/src/main/java/com/graphhopper/routing/util/AbstractFlagEncoder.java [hunk 1]
```diff
@@ -714,6 +725,7 @@ public double getDouble( long flags, int key )
         throw new UnsupportedOperationException("Unknown key " + key + " for double value.");
     }
 
+    @Deprecated
```
Comment (karussell) [inline core/src/main/java/com/graphhopper/routing/util/AbstractFlagEncoder.java]: that is indeed important :)
Would you also add this into the https://github.com/graphhopper/graphhopper/blob/master/core/files/changelog.txt and comment on the alternative?

File: core/src/main/java/com/graphhopper/storage/StorableProperties.java [hunk 2]
```diff
@@ -168,6 +168,9 @@ public boolean checkVersions( boolean silent )
         {
             return false;
         }
+
+        // The check for the encoder version is not done here as this class does not know about the registered encoders
+
```
Comment (karussell) [inline core/src/main/java/com/graphhopper/storage/StorableProperties.java]: maybe point to the place where it is?

File: core/src/main/java/com/graphhopper/util/PMap.java [hunk 3]
```diff
@@ -27,7 +27,7 @@
  */
 public class PMap
 {
-    private final Map<String, String> map;
+    protected final Map<String, String> map;
```
Comment (karussell) [inline core/src/main/java/com/graphhopper/util/PMap.java]: why do we need this protected?

File: core/src/test/java/com/graphhopper/routing/util/EncodingManagerTest.java [hunk 4]
```diff
@@ -66,6 +71,13 @@ public void testEncoderAcceptNoException()
     }
 
     @Test
+    public void testEncoderWithWrongVersionIsRejected()
+    {
+        thrown.expect(IllegalArgumentException.class);
```
Comment (karussell) [inline core/src/test/java/com/graphhopper/routing/util/EncodingManagerTest.java]: nice. I should have done this that way in lots of other places too :)

**Comment by karussell** [comment]
Looks good, had just very minor comments.

File: core/src/main/java/com/graphhopper/routing/util/AbstractFlagEncoder.java (hunk 1 above)
Comment (andreaswolf) [inline core/src/main/java/com/graphhopper/routing/util/AbstractFlagEncoder.java]: Will do, sure.

File: core/src/main/java/com/graphhopper/util/PMap.java (hunk 3 above)
Comment (andreaswolf) [inline core/src/main/java/com/graphhopper/util/PMap.java]: Not anymore, needed this during migration. Will roll back.

File: core/src/test/java/com/graphhopper/routing/util/EncodingManagerTest.java (hunk 4 above)
Comment (andreaswolf) [inline core/src/test/java/com/graphhopper/routing/util/EncodingManagerTest.java]: I can go through some existing tests and adjust them, would be a good way for me to learn more about JUnit…

Comment (karussell) [inline core/src/test/java/com/graphhopper/routing/util/EncodingManagerTest.java]: yeah, but not a high priority. maybe fix them when you stumble over them step by step :)

BTW: One more important issue for the testing stuff is a simple graph creation UI: https://github.com/graphhopper/graphhopper/issues/143

Comment (dos65) [inline core/src/test/java/com/graphhopper/routing/util/EncodingManagerTest.java]: Also there is another way to define expected exception in test
`@Test(expected = IllegalArgumentException.class)`

Comment (karussell) [inline core/src/test/java/com/graphhopper/routing/util/EncodingManagerTest.java]: Thanks, even better :) !

Comment (andreaswolf) [inline core/src/test/java/com/graphhopper/routing/util/EncodingManagerTest.java]: AFAIK there’s no way to test for the exception message in the annotation, right?

**Comment by andreaswolf** [comment]
First pushed my new three commits, after that a squashed version that contains all changes in one commit (and a minor unrelated change in a different commit).

File: core/src/test/java/com/graphhopper/routing/util/EncodingManagerTest.java (hunk 4 above)
Comment (karussell) [inline core/src/test/java/com/graphhopper/routing/util/EncodingManagerTest.java]: Probably but I don't know ... I was just used to catch+assert

**Comment by karussell** [comment]
Nice, thanks!

File: core/src/test/java/com/graphhopper/routing/util/EncodingManagerTest.java (hunk 4 above)
Comment (dos65) [inline core/src/test/java/com/graphhopper/routing/util/EncodingManagerTest.java]: Yes, there is no way to test exception message using annotation, but i think it's doesn't matter in tests like this :).

**Pull request merged**